teps/0121-unified-retries-for-taskruns-and-customruns.md

```diff
@@ -0,0 +1,152 @@
+
+### Goals
+1. Align the behavior of `Retries` between TaskRun and CustomRun.
+2. Retry TaskRun independently.
+
+### Non-Goals
+N/A
+
+## Related Work
+
+In the CI/CD industry, most systems support retry, some of them enables users to configure the retry strategy[^retry-strategy].
+
+Typically, a retry strategy needs to specify:
+1. When to retry
+2. The amount of attempts
+3. Actions to take after a failed attempt
+4. Timeout of each attempt
+
+
+|  | [Retry Step in GA](https://github.com/marketplace/actions/retry-step) | [Retry Action in GA](https://github.com/marketplace/actions/retry-action) | [Argo Workflow](https://argoproj.github.io/argo-workflows/retries/#retry-policies) | [GitLab](https://docs.gitlab.com/ee/ci/yaml/#retry) | [Ansible](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html#retrying-a-task-until-a-condition-is-met)| [Concourse](https://concourse-ci.org/attempts-step.html#attempts-step) |
+|:---|:---|:---|:---|:---|:---|:---|
+| **When to Retry** | configurable | on failure |configurable|configurable|[always retry, conditional stop](https://github.com/ansible/ansible/pull/76101) [^ansible-conditional-stop]|configurable|
+| **Attempts amount** |supported|supported|supported|supported|supported|supported|
+| **Timeout for each attempt** |supported|supported|supported|[supported](https://docs.gitlab.com/ee/ci/yaml/#retrywhen)|supported|supported|
+| **Timeout for all attempts** |[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[supported](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepstimeout-minutes)|[?](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout)|-|-|-|
+
+
+Several observations regarding of the feature table above:
+- We can configure **timeout duration per attempt** in all CI systems that support the `retry` functionality.
+- GitHub Action doesn't support retry natively, but because the flexibility of **customized actions**, some users write their own `retry` action to make it work, and those customized actions even support what to do before retrying a failed attempt.
+- Concourse mentioned the reason it [retries per attempt is somewhat arbitrary](https://concourse-ci.org/attempts-step.html#attempts-step).
+- Argo Workflow supports timeout per attempt, but [this stackoverflow question](https://stackoverflow.com/questions/71424429/argo-workflowtemplate-timeout) implies the need of "timeout for all retries".
+
+## Proposal
+We propose to support `retries` in TaskRuns, and solidify that `Timeout` is for each Retry attempt in CustomRun by updating [TEP-0069](https://github.com/tektoncd/community/blob/33ca1d5254a405b1d479f2350443f6c7979a0b72/teps/0069-support-retries-for-custom-task-in-a-pipeline.md) and the Developer guidance [docs/runs.md](https://github.com/tektoncd/pipeline/blob/0dd9f9719177054bdf60928ce9b4aab4b85af5bd/docs/runs.md). 
```

Comment:
Should we make this change to guidance for custom runs only for v1beta1.CustomRun, and not v1alpha1.Run? Or does it even matter, since we're just passing the timeout directly to the run? Maybe we can say that custom run controllers should interpret timeout as being for each individual attempt, and specify in their docs if this is not the case?

Reply:
That's a great question. I would prefer we need to clarify, and need to do it for both versions. 

It matters. We need to guide developer how to implement it because we need to build a protocol between pipelinerun controller and customrun controller, so that in pipelinerun controller, we are able to know when the Run finished, or should be retried, based on the `retries`  and `retriesStatus`.

The reason that we need to clarify it in v1alpha1.Run is our current [implementation](https://github.com/tektoncd/pipeline/blob/e1fd72720652fa5a1b7f811a6ce1d6c943be11a6/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L198) of the `retries` in v1alpha1.Run is different from what's documented as @jerop pointed out previously.

Reply:
I think you're right that it does matter. I think we might want to make the timeout change in v1beta1 only (but I could be convinced otherwise) because it will make migration easier-- custom run controllers will need to be updated to support v1beta1 customruns so changes can be made then.

re: guidance for retries status: maybe worth discussing in the other thread whether we truly need retriesstatus? Either way, I don't think our guidance for retries status is changing, just timeouts per retry, but please correct me if I'm wrong

Reply:
Created an issue https://github.com/tektoncd/pipeline/issues/5582 to track the effort around this behavior deviation.